### DISCUSSION TITLE
print doesn't come back after light sleep on rpi pico

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#14401] [RP2040 / Pico] [answered]
print seems problematic on the rpi pico (upython v 1.22.2)
```
from machine import lightsleep,Pin; from time import sleep
led = Pin(25, Pin.OUT)
for i in range(3):
  print('cycle', i); sleep(.1)
  led(1)
  lightsleep(1000)
  led(0)
  lightsleep(1000)
print ('end') 
>>> 
cycle 0
```
The led flashes 3 times which means the loop runs as expected but I get only one of the 4 prints I expected. Is there a work around for this behaviour?

**Comment by davefes** [comment]
Try adding:
`time.sleep_ms(5)`
after and before light and deepsleep and machine.reset() and

**Reply by kjm1102** [reply]
No joy Dave. What a bummer, feeling handicapped here with my fav diagnostic (the strategically placed print stmt) rendered inoperative.

**Reply by davefes** [reply]
Hmmm, been doing that for the last year or so on ESP32s.

**Comment by robert-hh** [comment]
This blocking of print is caused by lightsleep(). lightsleep turns of the USB clock and slows down other clocks. It seems that USB needs input from the PC to get working again. The same would happen if you use sys.stdout.write() instead of print(). Using sleep_ms() instead of lightsleep() works.

**Reply by kjm1102** [reply]
I need to lightsleep the pico to get the reduced power consumption Rob. I guess I could use sleep instead of lightsleep for program development then pull the diagnostic prints & swap to lightsleep when it's all working. 

Is there a cmd I could issue after lightsleep has expired to restart the USB clock?

**Reply by robert-hh** [reply]
> Is there a cmd I could issue after lightsleep has expired to restart the USB clock?

I do not know that. I just have seen these blocks after lightsleep(), and do not eve know whether the board or the PC is the problem. ~~Maybe the USB shut-off in lightsleep() has to be dropped.~~

**Reply by projectgus** [reply]
Do you see much difference in power consumption with sleep_ms vs lightsleep now that rp2 port is tickless?

(I think this is probably still a bug with lightsleep, but the alternative might not be as power hungry as it previously was.)

Note there is still a potential behaviour difference, as lightsleep may continue normal execution before the timeout period has elapsed, but sleep_ms won't.

**Reply by robert-hh** [reply]
> Do you see much difference in power consumption with sleep_ms vs lightsleep now that rp2 port is tickless?

lightsleep(): 1 mA
sleep_ms(): 19 mA

**Reply by projectgus** [reply]
Oof, fair enough!

**Reply by GitHubsSilverBullet** [reply]
Looks like ›lightsleep‹ breaks the USB connection completely.
```
[  775.956722] usb 1-1.3.4: new full-speed USB device number 17 using ehci-pci
[  776.067562] usb 1-1.3.4: New USB device found, idVendor=2e8a, idProduct=0005, bcdDevice= 1.00
[  776.067567] usb 1-1.3.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  776.067569] usb 1-1.3.4: Product: Board in FS mode
[  776.067571] usb 1-1.3.4: Manufacturer: MicroPython
[  776.067573] usb 1-1.3.4: SerialNumber:  xxx
[  776.068217] cdc_acm 1-1.3.4:1.0: ttyACM0: USB ACM device

[  786.231739] usb 1-1.3-port4: disabled by hub (EMI?), re-enabling...
[  786.232238] usb 1-1.3.4: USB disconnect, device number 17

[  786.466400] usb 1-1.3.4: new full-speed USB device number 18 using ehci-pci
[  786.784988] usb 1-1.3.4: New USB device found, idVendor=2e8a, idProduct=0005, bcdDevice= 1.00
[  786.784993] usb 1-1.3.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  786.784995] usb 1-1.3.4: Product: Board in FS mode
[  786.784997] usb 1-1.3.4: Manufacturer: MicroPython
[  786.784998] usb 1-1.3.4: SerialNumber:  xxx
[  786.785646] cdc_acm 1-1.3.4:1.0: ttyACM0: USB ACM device
```

**Reply by projectgus** [reply]
I see, that makes sense as there's nothing active in the device to process requests from the host, it'll trigger a bus reset instead.

The planned work to allow mpremote to wait for a USB device to come back (for dynamic USB support) should be able to help with this as well. That's provided mpremote is the program that's connected to MicroPython.

**Reply by kjm1102** [reply]
I'm surprised that lightsleep breaks the USB connection in that my windows PC is not emanating that characteristic sound it makes when I unplug something from the USB or try to use deepsleep on the rpi pico. 

I've just found a pico in the parts bin with I2C lcd, I'll use that for lightsleep diagnosis & quietly circumvent this can of worms.

**Reply by GitHubsSilverBullet** [reply]
It doesn't do it always, sometimes it just hangs and does not respond at all.

On a different note, I'm pretty sure that modifying mpremote wouldn't solve that problem.
I just wrote a small PC receiver (including trapping exceptions) and while the LED is still happily blinking, there's nothing coming in. Even restarting the PC part changes nothing, it's just timing out while waiting for any characters.

So  I'm rather confident that disabling the USB part (or its clocks) on the µC side isn't such a good idea after all. In all likelihood, USB needs to be fully reinitialized after a lightsleep in order to get it working properly.

If that means a complete disconnect/reconnect cycle, one has to keep in mind that the PC-OS keeps the old devicename for a second or two. In other words, disconnect/reconnect to fast and /dev/ttyACM0 will become /dev/ttyACM1.

**Reply by projectgus** [reply]
> On a different note, I'm pretty sure that modifying mpremote wouldn't solve that problem.

I need to implement this to be sure, but provided the USB device works correctly when it wakes up then it should give a pretty seamless experience. It's not trivial but it's possible, for example this is how the Espressif idf.py tool works when programming and debugging a device via the built-in microcontroller USB device interface (which also goes away over a reset cycle).

> So I'm rather confident that disabling the USB part (or its clocks) on the µC side isn't such a good idea after all.

I think there's a trade-off here, between having the lowest lightsleep consumption possible and allowing seamless USB debugging when coming in and out of lightsleep.

Although it may be sufficient to have logic to disable the USB device in lightsleep but only if it's not currently attached to a host. If there's a USB host connected then you can probably make the argument that a few mA extra power consumption aren't likely to matter.

**Reply by GitHubsSilverBullet** [reply]
I agree. There's always the possibility to attach a UART-to-USB converter and use that during debugging. I'd suggest to simply add a small note about possible problems to the ›lightsleep()‹ documentation.

**Reply by projectgus** [reply]
@GitHubsSilverBullet Do you recall how you triggered the kernel log above, where the host fully reset the rp2 in lightsleep? I wasn't able to reproduce, although I think it might come down to the exact USB host controller hardware.

**Reply by GitHubsSilverBullet** [reply]
@projectgus Unfortunately I was never able to ›realiably‹ provoke the crashes. Even a hardware reset followed by execution of the very same code behaved differently. But if `lightsleep()` was immediately followed by a `print()` it seemed to provoke the misbehavior more often.

On a different note, `lightsleep()` only makes sense if one needs to save even the tiniest bit of energy. USB on the other hand is the exact opposite, it provides energy to the controller.  Thus, while I surely appreciate the effort, I wouldn't waste any time on that matter. At all.

**Reply by projectgus** [reply]
> lightsleep() only makes sense if one needs to save even the tiniest bit of energy. USB on the other hand is the exact opposite, it provides energy to the controller. Thus, while I surely appreciate the effort, I wouldn't waste any time on that matter. At all.

I agree, but I think the use case that @kjm1102 has is a legitimate one (they want to debug their program which uses lightsleep by connecting it to a USB host). I've got a PR that should address this, though (see other reply).

**Reply by kjm1102** [reply]
Am I allowed to submit pull requests? I have another problem I'd like fixed https://github.com/orgs/micropython/discussions/14539

**Reply by projectgus** [reply]
@kjm1102 Anyone can submit a pull request! To increase the chance your pull request gets merged, suggest reading through some of the developer docs starting from https://github.com/micropython/micropython/wiki/ContributorGuidelines#sending-pull-requests

**Reply by kjm1102** [reply]
So what's the best way forward when I see a problem I think needs fixing but I don't have the smarts to offer the code to fix it Angus? https://github.com/orgs/micropython/discussions/14539

**Reply by projectgus** [reply]
@kjm1102 In the case of that discussion then I'd suggest adding more detail: exactly what hardware (development board, etc.) that you have, the exact code you're running that you expect to work, exactly what is connected to which pins on the board, etc etc.

It's almost impossible to provide too much detail, as long as the information is specific.

That will let someone try to reproduce the behaviour, or maybe even tell you why it doesn't work, without having to guess.

**Reply by kjm1102** [reply]
And I do that in a PR rather than here in the forum?

**Reply by projectgus** [reply]
I suggest adding that information in the Discussion you already created. A PR is for when you have code changes that you wish to contribute.

**Reply by kjm1102** [reply]
OK I expanded on my problem @ https://github.com/orgs/micropython/discussions/14539 but I'm not hopeful. It's not every day that an Angus comes along, sees a problem in a discussion, knows how to fix it then does a PR. Many thnx  for that BTW.

**Comment by projectgus** [comment]
I've submitted a pull request that keeps the USB peripheral clocks enabled during lightsleep on rp2, if a USB host is connected and active: https://github.com/micropython/micropython/pull/15111

**Reply by dpgeorge** [reply, marked as answer]
This PR was merged in a84c7a0ed931667cda44a8e27888f066e085d5eb (not part of the v1.23.0 release, will be in the next release).

**Comment by GitHubsSilverBullet** [comment]
Ahem, has this been tested?
```python
#!micropython
from machine import lightsleep, Pin
led = Pin(25, Pin.OUT)
for i in range(10):
    led.toggle()
    lightsleep(100)
for i in range(6):
    print(f'cycle {i}')
    led.toggle()
    lightsleep(500)
```

Hangs. Sometimes during the first run, sometimes during the second.
But it will hang. Not only the communication, the µC itself.
`MicroPython v1.23.0 on 2024-06-02; Raspberry Pi Pico with RP2040`

**EDIT:**
Interesting. Tried it on a different system (Notebook) using Debian 12 instead of Debian 11.
Was a bit more stable, took four runs until it crashed and hung. (USB2)
On the USB3 port it crashed almost always instantaneous.

**Reply by scruss** [reply]
can confirm intermittent lock-ups of a Pico W when run from mpremote. Curiously, indication that a crash is imminent seems to be that `cycle 0` is not echoed to the terminal, but 1–5 are.

Running it in Thonny is not recommended at all.

**Reply by projectgus** [reply]
The PR was merged after v1.23.0, it won't be in a release until v1.24. I'll update the other reply to make that clear.

If it's still not working on the latest v1.24 preview build then that's concerning, though. I wasn't able to reproduce the exact behaviour @GitHubsSilverBullet saw before the change, so it's possible there are host controllers that still don't like the way the RP2040 behaves after the change.

**Reply by projectgus** [reply]
Also, I appreciate that it's frustrating when things don't work as expected however I don't appreciate the tone of comments such as *"Ahem, has this been tested?"*. The PR description has a specific list of tests that were run, and explains in detail how there is some reported behaviour that I was unable to reproduce on my own hardware.

I'm happy to work with folks to debug this kind of stuff, but only if it's going to be in a mutually respectful way.

**Reply by GitHubsSilverBullet** [reply]
@projectgus It wasn't meant to be disrespectful at all. Sorry if it came over the wrong way. Sometimes the language barrier isn't helpful either. 
So, what I meant was: During a test I found that it's still not working and wanted to know whether the problem is somewhere on my side or if the fix is still in the making.
Oh, and I'm not frustrated at all. I wouldn't mind if it's left in its current state. Explained my reasoning before.

**Reply by projectgus** [reply]
No worries at all, I appreciate the explanation. If it's not working on v1.24-preview then please do let us know.

Are you both using Thonny? I don't know what happens if the host (i.e. Thonny) tries to send some data over USB while the RP2040 is in light sleep. It will be NAKed at the USB layer, and the host will retry, but maybe after a while the host still gives up and sends a write error back to Thonny.

**Reply by GitHubsSilverBullet** [reply]
That's good to hear. 
No, I have never used Thonny, I feel that it introduces far more problems than it promises to solve.
Vi, a slightly modified version of  pyboard and miniterm is all I need.
Will get the newest preview, give it a try and post what transpires.

**EDIT:**
OK, tested with:
```python
MPY: soft reboot
MicroPython v1.24.0-preview.23.g8809ae716 on 2024-06-05; Raspberry Pi Pico with RP2040
```
Here's a slightly modified python code to test:
```python
#!micropython
from machine import lightsleep, Pin
from time import sleep_ms

def flicker(delay=50,num=10):
    for _ in range(num):
        led.toggle()
        sleep_ms(delay)

led = Pin(25, Pin.OUT)
for i in range(10):
    flicker()
    print(f'cycle {i}')
    sleep_ms(50) # needed to print anything at all
    lightsleep(500)
flicker(25,50)
led.on()
```

dmesg showed neither errors nor anomalies.
```
[ 7526.144550] usb 1-1.3.4: new full-speed USB device number 23 using ehci-pci
[ 7526.255349] usb 1-1.3.4: New USB device found, idVendor=2e8a, idProduct=0005, bcdDevice= 1.00
[ 7526.255354] usb 1-1.3.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 7526.255365] usb 1-1.3.4: Product: Board in FS mode
[ 7526.255366] usb 1-1.3.4: Manufacturer: MicroPython
[ 7526.255368] usb 1-1.3.4: SerialNumber: ba4d9216214d8a53
[ 7526.257126] cdc_acm 1-1.3.4:1.0: ttyACM0: USB ACM device
```
On the µC the code always ran through to the end. (verified by observing the flicker patterns)
But only the first 2,3,4, sometimes 5 cycles were printed and it was not possible to get the REPL prompt afterwards.
While I was able to connect to the µC, I did not get a response. All the Ctrl-C…D…E inputs didn't bring it back to life. Needed a hard reset.

**Reply by projectgus** [reply]
Thanks @GitHubsSilverBullet for the clear explanation and test case.

I can reproduce this intermittently here. Most of the time it completes, but sometimes it gets stuck as described. Will investigate more.

**Reply by GitHubsSilverBullet** [reply]
@projectgus  It's been quite a while but I seem to remember that there are a couple of timeouts which have to be observed.
Had a look at the USB Spec Rev2.0 April 27, 2000 to verify.
Table 7-14. Device Event Timings (Continued) on page 189 shows a number of different 50ms timeouts as well as one 500ms timeout. I wonder whether that might be a potential source for complications.

**Reply by projectgus** [reply]
I think this is solved, fix and a new test case here: https://github.com/micropython/micropython/pull/15301

> Table 7-14. Device Event Timings (Continued) on page 189 shows a number of different 50ms timeouts as well as one 500ms timeout. I wonder whether that might be a potential source for complications.

Hmm, yeah, probably at the minimum there are device control transfers where NAKing them continually isn't allowed. I think this should be safe from the device's perspective though, either the USB host stack should return an error or it will keep trying.

It wasn't exactly what happened here though, it seems like reconfiguring the PLL on wake would sometimes introduce a clock glitch that broke USB entirely (no more responses at all, NAK or otherwise.)
